### DISCUSSION TITLE
Embeddings out of the box and other doubts...

### DISCUSSION THREAD

**Discussion by carlada-ita** [MaartenGr/BERTopic#1433] [General]
Hello,
"in primis" thank you for the great works that you are doing!

I'm testing BERTopic for a classification work on italian documents from a specific domain.
Using different BERTopic configuration I have some doubts...

1. are the two BERTopic pipelines following equivalent? Can I use ALL BERTopic methods with both pipelines?

BERTopic Model Complete

		embedding_model = SentenceTransformer("paraphrase-multilingual-mpnet-base-v2")
		umap_model = UMAP(...)
		hdbscan_model = HDBSCAN(...)
		vectorizer_model = CountVectorizer(....)
		ctfidf_model = ClassTfidfTransformer()
		representation_model = KeyBERTInspired()
		
		topic_model = BERTopic(
			embedding_model=embedding_model,  
			umap_model=umap_model
			hdbscan_model=hdbscan_model,  
			vectorizer_model=vectorizer_model,  
			ctfidf_model=ctfidf_model
			representation_model=representation_model,      
			calculate_probabilities=calculate_probabilities,
			language=language,
			top_n_words=top_n_words, 
			verbose=verbose
		)
		topics, probabilities = topic_model.fit_transform(dataset_training)

BERTopic with SentenceTransformer "out of the box"

		embedding_model = SentenceTransformer("paraphrase-multilingual-mpnet-base-v2")
		embeddings = embedding_model.encode(dataset_training)
		
		umap_model = UMAP(...)
		hdbscan_model = HDBSCAN(...)
		vectorizer_model = CountVectorizer(....)
		ctfidf_model = ClassTfidfTransformer()
		representation_model = KeyBERTInspired()

		topic_model_OOB = BERTopic(
			umap_model=umap_model
			hdbscan_model=hdbscan_model,  
			vectorizer_model=vectorizer_model,  
			ctfidf_model=ctfidf_model
			representation_model=representation_model,      
			calculate_probabilities=calculate_probabilities,
			language=language,
			top_n_words=top_n_words, 
			verbose=verbose
		)
		topics, probabilities = topic_model_OOB.fit_transform(dataset_training, embeddings )

2. If I want to execute the PREDICTION step with a new test file in BERTopic OOB, is it necessary to create a new embedding for the test file? Is it correct?

        embeddings_test = embedding_model.encode(dataset_test)
        topics_test, probs_test = topic_model_OOB.transform(dataset_test, embeddings_test)

3. After training the BERTopic model on dataset_training (es. topic_model.fit_transform(...)), I use reduce_outliers() for the redistribuition of elements in topics -1 and saved the new topics in the model

		new_topics = topic_model.reduce_outliers(dataset_training, topic_model.topics_[:], strategy="c-tf-idf")
		topic_model.update_topics(dataset_training, topics=new_topics)
	
	Then, I execute the PREDICTION step on dataset_test 
	
		topics_test, probs_test = topic_model.transform(dataset_test)
		
	I noticed that the topics in topics_test are the same of those in the training step...I assumed they could be the same of topics_test instead...Is it an incorrect assumption? Is it necessary to use reduce_outliers() also on topics_test?
	
Thank you in advance for your attention.

Carlo

**Comment by MaartenGr** [comment]
> are the two BERTopic pipelines following equivalent? Can I use ALL BERTopic methods with both pipelines?

Not exactly, you will need to specify `embedding_model=embedding_model` in the second pipeline for them to be the same. 

> If I want to execute the PREDICTION step with a new test file in BERTopic OOB, is it necessary to create a new embedding for the test file? Is it correct?

You can do it the way you demonstrated if you wan to pre-calculate the embeddings. You can also skip pre-calculating the embeddings as that can also be done internally. 

> After training the BERTopic model on dataset_training (es. topic_model.fit_transform(...)), I use reduce_outliers() for the redistribuition of elements in topics -1 and saved the new topics in the model

The predictions will stay the same regardless of whether you are using outlier reduction beforehand since outlier reduction does not change the internal cluster model/assignment. You would indeed have to apply it afterwards.

**Reply by carlada-ita** [reply]
Hi Maarten,
Thank you  for your answers.

Some considerations...

> _Not exactly, you will need to specify embedding_model=embedding_model in the second pipeline for them to be the same._

Do you refers to -->?
```
        embedding_model = SentenceTransformer("paraphrase-multilingual-mpnet-base-v2")
	embeddings = embedding_model.encode(dataset_training)
	
        umap_model = UMAP(...)
	hdbscan_model = HDBSCAN(...)
	vectorizer_model = CountVectorizer(....)
	ctfidf_model = ClassTfidfTransformer()
	representation_model = KeyBERTInspired()

	topic_model_OOB = BERTopic(
               -->      embedding_model=embedding_model
		umap_model=umap_model
		hdbscan_model=hdbscan_model,  
		vectorizer_model=vectorizer_model,  
		ctfidf_model=ctfidf_model
		representation_model=representation_model,      
		calculate_probabilities=calculate_probabilities,
		language=language,
		top_n_words=top_n_words, 
		verbose=verbose
	)
	topics, probabilities = topic_model_OOB.fit_transform(dataset_training, embeddings )
```

But in this case SentenceTransformer will also be processed by BERTopic model again...is it correct? If so, why calculate embeddings "out of the box"? what am I missing?

> You can do it the way you demonstrated if you wan to pre-calculate the embeddings. You can also skip pre-calculating the embeddings as that can also be done internally.

How can I invoke transform method without pre-calculated embeddings (in the case of "out of the box" embeddings)? 

`topics_test, probs_test = topic_model_OOB.transform(dataset_test, ??????)`

Please, can you give me an example?


> The predictions will stay the same regardless of whether you are using outlier reduction beforehand since outlier reduction does not change the internal cluster model/assignment. You would indeed have to apply it afterwards.

Therefore, if I understand correctly, the reduce_outliers() method "must be called" on  each different dataset (training, test, etc.)


Thank you so much.

**Comment by MaartenGr** [comment]
> But in this case SentenceTransformer will also be processed by BERTopic model again...is it correct? If so, why calculate embeddings "out of the box"? what am I missing?

If you do not specify an `embedding_model` then the model will not know which embedding model to use for `KeyBERTInspired`. Even if you pass the embeddings directly, adding `embedding_model` will help in a number of tasks, including `.transform`. 

> How can I invoke transform method without pre-calculated embeddings (in the case of "out of the box" embeddings)?

By doing this:

```python
topics_test, probs_test = topic_model_OOB.transform(dataset_test)
```

> Therefore, if I understand correctly, the reduce_outliers() method "must be called" on each different dataset (training, test, etc.)

Yes, the `reduce_outliers` does not change the internal model so you would have to run that each time on a different dataset.
